### PR TITLE
Companies Generation & Creation

### PR DESCRIPTION
**Pull request by kushniryb**
Hey
This is JS part of the WT App

### CODE REVIEW

**Comment by beornborn** [comment]
plagiarism

**Comment by kushniryb** [comment]
What are you talking about?

**Comment by beornborn** [comment]
very similar to https://github.com/WatchTowerBenefits/wt_javascript_test/pull/11/files

**Comment by kushniryb** [comment]
Yeah, I see what you mean now, there is some similarity in terms of naming/architecture but that's just a common sense. 

Obviously it's up to you but today was the first time I've taken a look at that PR. It would be really cool if you could still review it. I've put an effort in it and it would be a shame not to get a chance for a reason like this. Please, take a look at my rails API / components structure, maybe that'll prove my point 

Best regards, Yura

**Comment by beornborn** [comment]
@kushniryb ok, I'll review your task

**Comment by kushniryb** [comment]
@beornborn just curious, when will you have a chance to review this / Rails parts? It's been almost a week :|

**Comment by beornborn** [comment]
@kushniryb yes I know, sorry about that, really tough week

**Comment by kushniryb** [comment]
Ok, thank you @beornborn

File: app/Root.js [hunk 1]
```diff
@@ -1,23 +1,33 @@
...
-        </Route>
-      </Router>
-    </div>
+    <MuiThemeProvider>
+      <div style={{height: '100%'}}>
+        <Router history={browserHistory}>
+          <Redirect from='/' to='/companies' />
+          <Route path='/' component={Layout}>
+            <Router path='/companies' component={Companies} />
+            <Router path='/companies/new' component={CreateCompany} />
```
Comment (beornborn) [inline app/Root.js]: better NewCompany

File: app/components/pages/companies/CompaniesLayout.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    headerTitle: pt.string.isRequired,
+    children: pt.node
+  },
+
+  render() {
+    const { headerTitle } = this.props
+
+    return <div>
+      <AppBar title={headerTitle} />
+      <Container>{this.props.children}</Container>
```
Comment (beornborn) [inline app/components/pages/companies/CompaniesLayout.js]: why not `const { headerTitle, children } = this.props` ?

File: app/components/pages/companies/CompaniesTable.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+const CompaniesTable = React.createClass({
+  propTypes: {
+    companies: pt.arrayOf(pt.object).isRequired
+  },
+
+  render() {
+    const { companies } = this.props
+
+    return <Table selectable={false}>
```
Comment (beornborn) [inline app/components/pages/companies/CompaniesTable.js]: why do we see checkboxes?

File: app/components/pages/companies/CompanyForm.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    return <form onSubmit={handleSubmit}>
+      <div>
+        <Field name='name' component={TextField} hintText='Company Name' />
+      </div>
+      <RaisedButton secondary label='Create Company' type='submit' />
+    </form>
+  }
+})
+
+export default reduxForm({
```
Comment (beornborn) [inline app/components/pages/companies/CompanyForm.js]: better keep this in container

File: app/containers/pages/companies/Companies.js [hunk 5]
```diff
@@ -1,15 +1,17 @@
...
+import Companies from '~/app/components/pages/companies/Companies'
 
 export const mapStateToProps = (state: Object): Object => ({
   companies: state.company.companies,
+  pagination: state.company.pagination
 })
 
 export const mapDispatchToProps = (dispatch: Function): Object => ({
-  getCompanies: () => dispatch(createAction(SAGA_GET_COMPANIES)())
+  getCompanies: (page: number = 1) => dispatch(createAction(SAGA_GET_COMPANIES)(page)),
```
Comment (beornborn) [inline app/containers/pages/companies/Companies.js]: better {page}

File: app/containers/pages/companies/CreateCompany.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
+// @flow
+import { connect } from 'react-redux'
+import { createAction } from 'redux-actions'
+import { SAGA_CREATE_COMPANY } from '~/app/reducers/Company'
+import CreateCompany from '~/app/components/pages/companies/CreateCompany'
+
+export const mapStateToProps = (_state: Object): Object => ({})
+
+export const mapDispatchToProps = (dispatch: Function): Object => ({
+  createCompany: (values) => dispatch(createAction(SAGA_CREATE_COMPANY)({company: values}))
```
Comment (beornborn) [inline app/containers/pages/companies/CreateCompany.js]: flow type for values?

File: app/components/pages/companies/Companies.js [hunk 7]
```diff
@@ -1,26 +1,41 @@
...
 
-    return <Container>
-      <Header>Companies</Header>
-      {companies.map(c =>
-        <Company key={c.id}>{c.name}</Company>
-      )}
-    </Container>
+    return <CompaniesLayout headerTitle='Companies'>
+      <CardActions style={{display: 'inline-flex'}}>
+        <RaisedButton primary label='Create Company' href='/companies/new' />
```
Comment (beornborn) [inline app/components/pages/companies/Companies.js]: it does full page reload

File: package.json [hunk 8]
```diff
@@ -17,11 +17,14 @@
...
     "immutability-helper": "^2.1.1",
     "lodash": "^4.17.4",
+    "material-ui": "^0.17.1",
+    "material-ui-pagination": "^1.0.1",
     "moment": "^2.17.1",
     "normalizr": "^3.2.1",
     "react": "^15.4.2",
     "react-bootstrap": "^0.30.7",
     "react-dom": "^15.4.2",
+    "react-grid-system": "^2.7.0",
```
Comment (beornborn) [inline package.json]: don't see it's used

File: app/reducers/Company.js [hunk 9]
```diff
@@ -4,16 +4,28 @@ import update from 'immutability-helper'
...
 
 const initialState = {
-  companies: []
+  companies: [],
+  pagination: {
+    current: 1,
+    total: 0,
+    display: 30
+  },
+  name: ''
```
Comment (beornborn) [inline app/reducers/Company.js]: what is this?

Comment (beornborn) [inline app/reducers/Company.js]: eslint returns error

File: app/components/pages/companies/CompanyForm.js (hunk 4 above)
Comment (kushniryb) [inline app/components/pages/companies/CompanyForm.js]: Could you please elaborate on this?

Comment (beornborn) [inline app/components/pages/companies/CompanyForm.js]: example

```
// @flow
import { connect } from 'react-redux'
import Form from '~/app/components/dialogs/add_new_member/Form'
import { createAction } from 'redux-actions'
import { SAGA_CREATE_MEMBER } from '~/app/reducers/MemberList'
import { SAGA_GET_HONORIFICS } from '~/app/reducers/Common'
import { reduxForm, formValueSelector } from 'redux-form'

export const form = {
  form: 'member/add',
  validate: (values: Object) => {
    if (!values) return {}
    const errors = {}
    if (!values.email) {
      errors.email = 'Required'
    } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) {
      errors.email = 'Invalid email address'
    }
    if (!values.first_name) { errors.first_name = 'Required' }
    if (!values.last_name) { errors.last_name = 'Required' }
    return errors
  }
}

export const mapStateToProps = (state: Object): Object => {
  const formSelector = formValueSelector('member/add')

  return {
    referral_type: formSelector(state, 'referral_type'),
    honorifics: state.common.honorifics,
    initialValues: {
      birthday: '',
      primary_address: {
        country_code: '',
        locality: '',
        postal_code: '',
        region: '',
        street_address: ''
      },
      mobile_phone: {country_calling_code: 1, number: ''}
    }
  }
}

export const mapDispatchToProps = (dispatch: Function): Object => ({
  onSubmit: (formData: Object) =>
    new Promise((resolve, reject) => {
      dispatch(createAction(SAGA_CREATE_MEMBER)({formData, resolve, reject}))
    }),
  getHonorifics: () => dispatch(createAction(SAGA_GET_HONORIFICS)())
})

export default connect(mapStateToProps, mapDispatchToProps)(reduxForm(form)(Form))
```

Comment (beornborn) [inline app/components/pages/companies/CompanyForm.js]: components should contain less logic as possible, just presentation

Comment (kushniryb) [inline app/components/pages/companies/CompanyForm.js]: Oh, I see
Thank you, that makes much more sense now :+1:

**Pull request closed without merging**